### PR TITLE
Generalising TabulatedFluidProperties

### LINKED ISSUE #11084: Generalise TabulatedFluidProperties
## Rationale

I've had a request to make `TabulatedFluidProperties` more general by allowing more properties to be read from a table of data. I think that the use case is for including hydrocarbons, where it is apparently common to define their properties in a table of data.

## Description

`TabulatedFluidProperties` currently uses interpolation to calculate density, enthalpy and internal energy only, with all other fluid properties calculated using a supplied userobject. This was done as these properties were expensive to calculate for CO2 (as they had to be calculated iteratively).

There's no technical reason why `TabulatedFluidProperties` can't be generalised to use interpolation for more properties, and now that it is much, much faster than before (see #11060), perhaps using interpolation may be faster for many fluid properties (it at least won't be much slower).

## Impact

This will make it possible to define a fluid using only a table of data, and allow users to use interpolation for other fluid properties (and not just density, enthalpy and internal energy). Otherwise, there should be no change to the existing use of this class.

### PR DESCRIPTION
**Pull request by cpgr**
Closes #11084

### CODE REVIEW

File: modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h [hunk 1]
```diff
@@ -228,8 +218,34 @@ class TabulatedFluidProperties : public SinglePhaseFluidPropertiesPT
...
+  std::vector<std::string> _read_properties;
+  /// Set of flags to note whether a property is to be interpolated
+  bool _interpolate_density = false;
+  bool _interpolate_enthalpy = false;
+  bool _interpolate_internal_energy = false;
+  bool _interpolate_viscosity = false;
+  bool _interpolate_k = false;
+  bool _interpolate_cp = false;
+  bool _interpolate_cv = false;
+  bool _interpolate_entropy = false;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h]: While setting the initial values in a header file is valid C++11, I would move this in the c-tor initializer list, just for consistency.

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 2]
```diff
@@ -100,11 +108,76 @@ TabulatedFluidProperties::initialSetup()
...
+      // Only check properties not in _required_columns
+      if (std::find(_required_columns.begin(), _required_columns.end(), column_names[i]) ==
+          _required_columns.end())
+      {
+        if (std::find(_property_columns.begin(), _property_columns.end(), column_names[i]) ==
+            _property_columns.end())
+          mooseError(column_names[i],
+                     " read in ",
+                     _file_name,
+                     " is not one of the properties that TabulatedFluidProperties understands");
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C]: In general, I really like to see the name of the object that triggered the error. It helps to track down the error in the input file.  You probably do not have heaps of fluid properties in yours, but I still think it is a good idea.

You would change this like so, if you decided to change this:
```
mooseError(name(), "your error message");
```

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 3]
```diff
@@ -100,11 +108,76 @@ TabulatedFluidProperties::initialSetup()
...
+        else
+          _read_properties.push_back(column_names[i]);
+      }
+    }
+
+    // At this point, all properties read are able to be used by TabulatedFluidProperties.
+    // Now set flags and indexes for each propery to use in property calculations
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C]: ...each proper**t**y to use...

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h (hunk 1 above)
Comment (cpgr) [reply modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h]: I'll change this

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C (hunk 2 above)
Comment (cpgr) [reply modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C]: And this

File: modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h [hunk 4]
```diff
@@ -228,8 +218,34 @@ class TabulatedFluidProperties : public SinglePhaseFluidPropertiesPT
...
+  /// Index of each property
+  unsigned int _density_idx;
+  unsigned int _enthalpy_idx;
+  unsigned int _internal_energy_idx;
+  unsigned int _viscosity_idx;
+  unsigned int _k_idx;
+  unsigned int _cp_idx;
+  unsigned int _cv_idx;
+  unsigned int _entropy_idx;
+
```
Comment (cpgr) [inline modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h]: These are only set if the property is read from the file / specified in the input, so some of them might never be initialised. Should I give them some default value?

Comment (andrsd) [reply modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h]: That would not hurt anything

**Review by andrsd — APPROVED** [review]

**Pull request merged**